### PR TITLE
fix(auth): Vector does not put the Proxy-Authorization header on the wire (#17353)

### PR DESCRIPTION
**Pull request by syedriko**
fix(auth): Vector does not put the Proxy-Authorization header on the wire (#17353) 

closes: #17353

### CODE REVIEW

**Comment by StephenWakely** [comment]
I can see this is coming from the [`hyper-proxy`](https://docs.rs/hyper-proxy/latest/src/hyper_proxy/lib.rs.html#196-210) crate, but it isn't clear to me why we would want to set both `Authentication` and `Proxy-Authentication` headers. It strikes me as incorrect, but I am probably misunderstanding something.

Do you have any ideas?

**Comment by syedriko** [comment]
> I can see this is coming from the [`hyper-proxy`](https://docs.rs/hyper-proxy/latest/src/hyper_proxy/lib.rs.html#196-210) crate, but it isn't clear to me why we would want to set both `Authentication` and `Proxy-Authentication` headers. It strikes me as incorrect, but I am probably misunderstanding something.
> 
> Do you have any ideas?

It's rather Authorization and Proxy-Authorization. I don't think it's outright incorrect, as Proxy-Authorization stops at the proxy and Authorization gets passed on to the target, and there are situations where this is appropriate, but it's lacking nuance. I don't follow the logic of https://docs.rs/hyper-proxy/0.9.1/src/hyper_proxy/lib.rs.html#196-210 very well either. Setting both headers to the same credentials, as well as HTTP -> Authorization, HTTPS -> Proxy-Authorization... But sending out no headers is not good either.

File: src/http.rs [hunk 1]
```diff
@@ -135,6 +142,17 @@ where
 
         Box::pin(fut)
     }
+
+    fn maybe_add_proxy_headers(&self, request: &mut Request<B>) {
+        if let Some(proxy_headers) = self.proxy_connector.http_headers(request.uri()) {
+            for (k, v) in proxy_headers {
+                let request_headers = request.headers_mut();
```
Comment (bruceg) [inline src/http.rs:149]: I suggest moving this out of the loop, to avoid calling it repeatedly.

File: src/http.rs [hunk 2]
```diff
@@ -135,6 +142,17 @@ where
...
+                if !request_headers.contains_key(k) {
```
Comment (bruceg) [inline src/http.rs:150]: Do we need to check if it exists, or can we just insert the new header? HTTP headers can contain multiple values.

**Review by bruceg — COMMENTED** [review]
I have a suggestion and a question below, but this looks like the right thing to do according to my reading of the [ProxyConnector](https://docs.rs/hyper-proxy/latest/hyper_proxy/struct.ProxyConnector.html#method.http_headers) documentation. Thanks for the contribution!

File: src/http.rs (hunk 1 above)
Comment (syedriko) [reply src/http.rs:149]: Absolutely, my oversight.

File: src/http.rs (hunk 2 above)
Comment (syedriko) [reply src/http.rs:150]: I'm going by RFC 7235, https://github.com/hyperium/headers/blob/f01cc90cf8d601a716856bc9d29f47df92b779e4/src/common/authorization.rs#L9 and https://github.com/hyperium/headers/blob/master/src/common/proxy_authorization.rs#L3. My read of RFC 7235 is that it doesn't leave room for >1 sets of credentials in the Authorization and Proxy-Authorization headers.
The reason I added the existence check - https://github.com/vectordotdev/vector/blob/53e178570b5b87bc2124f4299865cbb00916fe20/src/http.rs#L293 can also set the Authorization header, which I didn't want to overwrite.

**Review by bruceg — APPROVED** [review]

**Pull request merged**